designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,544 @@
+1. Overriding of configuration via `overrides`
+
+All of these complexities arise from `.eslintrc` format of describing what should happen rather than how it should happen. We keep trying to anticipate how users want things to happen and guessing wrong has led to increased complexity.
+
+This RFC proposes a fresh start for configuration in ESLint that takes into account all of the requests we've received over the years. By simplifying and stripping configuration we can actually make a configuration system that is flexible enough to accomodate changes in the future and easy enough to implement that we won't be afraid to do so.
+
+## Detailed Design
+
+Design Summary:
+
+1. Introduce a new `.eslint.js` configuration file format
+1. Searching for `.eslint.js` starts from the current working directory and continues up
+1. All `.eslint.js` files are treated as if they have `root: true`
+1. There is no automatic merging of config files
+1. The `.eslint.js` configuration is not serializable and objects such as functions and plugins may be added directly into configuration
+1. Remove the concept of environments (`env`)
+1. Remove `.eslintignore` and `--ignore-file` (no longer necessary)
+1. Remove `--rulesdir`
+1. Create a `@eslint/config` package to help users merge configs
+
+### The `.eslint.js` File
+
+The `.eslint.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports a `config` object: 
+
+```js
+exports.config = {
+    files: "*.js",
+    ignore: "*.test.js",
+    globals: {},
+    settings: {},
+    processor: object,
+    parser: object,
+    parserOptions: {},
+    ruledefs: {},
+    rules: {}
+};
+```
+
+The following keys are new to the `.eslint.js` format:
+
+* `files` - **Required.** Determines the global pattern that this configuration applies to and also is used as the default pattern to search for when ESLint is executed on the command line.
+* `ignore` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this glob pattern are subtracted from the files specified in `files`.
+* `ruledefs` - Contains definitions for rules grouped by a specific name. This replaces the `plugins` key in `.eslintrc` files and the `--rulesdir` option.
+
+The following keys are specified the same as in `.eslintrc` files:
+
+* `globals`
+* `settings`
+* `rules`
+* `parserOptions`
+
+The following keys are specified differently than in `.eslintrc` files:
+
+* `parser` - an object in `.eslint.js` files (a string in `.eslintrc`)
+* `processor` - an object in `.eslint.js` files (a string in `.eslintrc`)
+
+Each of these keys used to require one or more strings specifying module(s) to load in `.eslintrc`. In `.eslint.js`, these are all objects, requiring users to manually specify the objects to use.
+
+The following keys are invalid in `.eslint.js`:
+
+* `env` - responsibility of the user
+* `extends` - responsibility of the user
```

Comment:
I think removing this, significantly complicates what I think is the most common scenario: I want to use AirBnB config, but I don't like 2-3 rules in it, and want to change their values. Before - I could just extend AirBnB and specify new config for 3 rules I didn't like. Now I need to require AirBnB config, retrieve rules object and override in place 3 rules I don't like. While it's not a huge difference, it does make it a bit more complicated, and if my assumption is correct (that it's the most common scenario) it will add a huge number of combined hours to creating basic configuration when adding up the time across all of the users.

Reply:
I see that this is addressed by `@eslint/config` package. Not sure that's simple enough to explain it to somebody without showing them the code.

Reply:
Yeah, the `@eslint/config` package is intended to solve this problem. 

For new projects, I envision using `--init` to setup a basic config that uses `Config.create()` by default to setup a file that is easier to modify.

Reply:
`extends` has been added back into this proposal.